### DISCUSSION TITLE
Change auto-import and type-checking behavior in conformance with Pep 585

### DISCUSSION THREAD

**Discussion by rben01** [microsoft/pylance-release#2637] [Enhancement]
[Pep 585](https://www.python.org/dev/peps/pep-0585/) says that in Python 3.9, use of the `typing` module's generics such as `Iterable`, `Generator`, etc. is deprecated as those types will be merged with the types of the same names in `collections.abc`. In addition, `typing`'s generic types that mimic built-in container types, such as `List` for `list` and `Tuple` for `tuple`, are similarly deprecated in favor of using the built-in types directly, e.g., `List[int]` should be written as `list[int]` going forward. In Python 3.7 and 3.8, these two behaviors can be achieved with `from __future__ import annotations`.

Ideally, in Python 3.9+ or in a 3.7 or 3.8 file containing that `__future__` import, Pylance would 1. auto-import the non-built-in types (`Iterable`, `Generator`, etc) from `collections.abc` instead of from `typing` 2. issue a warning when the user imports those types from `typing` instead of `collections.abc` 3. issue a warning if the user uses e.g., `typing.List` instead of `list`.

**Comment by judej** [comment]
Thanks for the report @rben01, will add this to our proposed enhancements.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by parched** [comment]
Currently, `collections.abc` aren't even on the list of suggestions, should that be considered a bug?
This is particularly annoying when you want to use something from `collections.abc` that isn't in `typing`.

![image](https://user-images.githubusercontent.com/5975405/183332441-65fa7898-c93e-46f0-a7d3-e70f9b0eb756.png)
